src/onnx/parse_matmul.cpp

```diff
@@ -89,55 +182,84 @@ struct parse_matmul : op_parser<parse_matmul>
+
+            if(is_quant_dot and
+               (not contains(supported_types, ba0_type) or
+                not contains(supported_types, ba1_type) or not contains(supported_types, a0_type) or
+                not contains(supported_types, a1_type)))
             {
-                auto l0_it = s0_lens.begin() + s0_lens.size() - 2;
-                std::vector<std::size_t> l0_broadcasted_lens(s0_lens.begin(), l0_it);
-                auto l1_it = s1_lens.begin() + s1_lens.size() - 2;
-                std::vector<std::size_t> l1_broadcasted_lens(s1_lens.begin(), l1_it);
-                auto output_lens =
-                    compute_broadcasted_lens(l0_broadcasted_lens, l1_broadcasted_lens);
-                l0_broadcasted_lens = output_lens;
-                l0_broadcasted_lens.insert(l0_broadcasted_lens.end(), l0_it, s0_lens.end());
-                l1_broadcasted_lens = output_lens;
-                l1_broadcasted_lens.insert(l1_broadcasted_lens.end(), l1_it, s1_lens.end());
-                if(s0_lens != l0_broadcasted_lens)
+                MIGRAPHX_THROW("PARSE_MATMULINTEGER: Unsupported type");
+            }
+
+            if(is_quant_dot)
+            {
+                instruction_ref offset_op;
+                if(a0_type == migraphx::shape::uint8_type or a1_type == migraphx::shape::uint8_type)
                 {
-                    ba0 = info.add_instruction(
-                        make_op("multibroadcast", {{"out_lens", l0_broadcasted_lens}}), a0);
+                    offset_op = info.add_literal(
+                        migraphx::literal{migraphx::shape{migraphx::shape::int16_type}, {-128}});
                 }
-                if(s1_lens != l1_broadcasted_lens)
+
+                // always convert uint8 to int8 to avoid rollover
+                if(a0_type == migraphx::shape::uint8_type)
                 {
-                    ba1 = info.add_instruction(
-                        make_op("multibroadcast", {{"out_lens", l1_broadcasted_lens}}), a1);
+                    a0 = add_int8_shift(info, offset_op, a0);
+                    if(has_ba0)
+                    {
+                        ba0 = add_int8_shift(info, offset_op, ba0);
+                    }
+                    else
+                    {
+                        ba0 = a0;
+                    }
                 }
-            }
 
-            // parse a_zero_point and b_zero_point values
-            if(args.size() > 2)
-            {
-                ba0 = info.add_instruction(
-                    make_op("convert", {{"target_type", migraphx::shape::float_type}}), ba0);
+                if(a1_type == migraphx::shape::uint8_type)
+                {
+                    a1 = add_int8_shift(info, offset_op, a1);
+                    if(has_ba1)
+                    {
+                        ba1 = add_int8_shift(info, offset_op, ba1);
+                    }
+                    else
+                    {
+                        ba1 = a1;
```

Comment:
Else is unnecessary `ba1 = a1` is there is no bias. It would be set by `set_bias_arg`

Reply:
issue is we still need to add the shift so without this we're using the unshifted version of a0/a1

Reply:
That would be taken care by `has_ba1` `has_ba2`

Reply:
but a0/a1 get shifted and then ba0/ba1 gets used for the appropriate dot operator to be used